File: longest.py
# Time Complexity - O(N) where N = no.of characters
# Space Complexity - O(N) where N = no.of characters
# Approach : Sliding window 

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        if not s:
            return 0

        index = 0
        max_num = float("-inf")

        hashmap = {}

        for idx, char in enumerate(s):
            if char in hashmap:

                index = max(index, hashmap[char]) 

            hashmap[char] = idx + 1 
            max_num = max(max_num, idx - index +1)

        return max_num

File: Longest.java
//Time Complexity: o(2n)
//Space Complexity: o(n)
/*Expln: sliding window over the string and record the longest substring 
 *by skipping over to the next char when you hit duplicates.
 */
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int val = 0;
        if(s.length() ==0) return val;
        HashMap<Character, Integer> map = new HashMap<>();
        int i = 0;
        int j = 1;
        map.put(s.charAt(i), i);
        val = j - i;
        for(; j<s.length(); j++)
        {
            if(map.containsKey(s.charAt(j)))
            {
                if(i <= map.get(s.charAt(j))){
                    i = map.get(s.charAt(j)) + 1;
                }
            }
            val = Math.max((j -i+1), val);
            map.put(s.charAt(j), j);
        }
        return val;
    }
}